README.md
# scrolling_day_calendar

A flutter calendar package to allow users to scroll through given dates either by 
swiping left and right or pressing the arrows. 
The active date is displayed as a heading and the tasks for that date are displayed underneath. 
This package can be used on Android and IOS

## How it works

* Pass in the widgets method
You simply need to pass in a start date, an end date, the active date and widgets to display per day.

* SetState method
You simply need to pass in a start date, an end date, the active date and a call-back function, 
the package will then allow the user to swipe between the dates using PageView, 
on each page change the call-back you have set will be called to allow you to update the page content for the given date.


## Screenshots
<img src="https://github.com/bmb-enterprises/scrolling-day-calendar/blob/master/sample_images/1.png?raw=true" height="300em" />
<img src="https://github.com/bmb-enterprises/scrolling-day-calendar/blob/master/sample_images/2.png?raw=true" height="300em" />


## Usage 

To use the plugin: 

* add the dependency to your [pubspec.yaml](https://github.com/bmb-enterprises/scrolling-day-calendar/tree/master/example)

```yaml
dependencies:
  flutter:
    sdk: flutter

  scrolling_day_calendar: 2.0.1

```

* import the package 
```dart 
import 'package:scrolling_day_calendar/scrolling_day_calendar.dart';

```

## How to use the Widget pass-in method - recommended for better experience
```dart
DateTime selectedDate = DateTime.now();
DateTime startDate = DateTime.now().subtract(Duration(days: 10));
DateTime endDate = DateTime.now().add(Duration(days: 10));
Map<String, Widget> widgets = Map();
String widgetKeyFormat = "yyyy-MM-dd";

body: ScrollingDayCalendar(
          startDate: startDate,
          endDate: endDate,
          selectedDate: selectedDate,
          dateStyle: TextStyle(
            fontWeight: FontWeight.bold,
            color: Colors.white,
          ),
          displayDateFormat: "dd, MMM yyyy",
          dateBackgroundColor: Colors.grey,
          forwardIcon: Icons.arrow_forward,
          backwardIcon: Icons.arrow_back,
          pageChangeDuration: Duration(
            milliseconds: 700,
          ),
          widgets: widgets,
          widgetKeyFormat: widgetKeyFormat,
          noItemsWidget: Center(
            child: Text("No items have been added for this date"), // add buttons etc here to add new items for date
          ),
        ),
```


## How to use the SetState method

```dart
// set the initial page value
Widget _pageItems = Text("Inital value");
DateTime selectedDate = DateTime.now();
DateTime startDate = DateTime.now().subtract(Duration(days: 10));
DateTime endDate = DateTime.now().add(Duration(days: 10));
String widgetKeyFormat = "yyyy-MM-dd";

// add to body of a Scaffold
body: ScrollingDayCalendar(
          startDate: startDate,
          endDate: endDate,
          selectedDate: selectedDate,
          onDateChange: (direction, DateTime selectedDate) {
            setState(() {
              pageItems = _widgetBuilder(selectedDate);
            });
          },
          dateStyle: TextStyle(
            fontWeight: FontWeight.bold,
            color: Colors.white,
          ),
          pageItems: pageItems,
          displayDateFormat: "dd/MM/yyyy",
          dateBackgroundColor: Colors.grey,
          forwardIcon: Icons.arrow_forward,
          backwardIcon: Icons.arrow_back,
          pageChangeDuration: Duration(
            milliseconds: 400,
          ),
          noItemsWidget: Center(
            child: Text("No items have been added for this date"), // add buttons etc here to add new items for date
          ),
        ),
```

CHANGELOG.md
## [0.0.1] - Release.

* A Customizable page scroll with calendar date header written in dart 2.

## [0.0.2] - Release.

* Fix screenshots.

## [1.0.1] - Release.

* Add more notes to read me, and also release the first version.

## [1.0.2] - Release.

* Fix screenshots.

## [1.0.3] - Release.

* Update example code to add date examples.

## [1.0.4] - Release.

* Remove the unnecessary state update.

## [1.0.5] - Release.

* Update site url

## [2.0.1] - Release.

* Add a new method to build pages by passing in widgets sorted by date, this allows for a more smoother scrolling

## [2.0.2] - Release.

* Add github link 

example/lib/test_data.dart
class TestData {
  static Map getTestData() {
    return {
      "2019-08-09": [
        {
          "title": "5th Aug ",
          "description":
              "sed ullamcorper morbi tincidunt ornare. Feugiat vivamus at augue eget arcu. Arcu cursus euismod",
        },
        {
          "title": "ad minim veniam, qu",
          "description": "ue. Lectus vestibulum mattis ullamcorper velit ",
        }
      ],
      "2019-08-16": [
        {
          "title": "16th Aug ",
          "description":
              "sed ullamcorper morbi tincidunt ornare. Feugiat vivamus at augue eget arcu. Arcu cursus euismod",
        },
        {
          "title": "ad minim veniam, qu",
          "description": "ue. Lectus vestibulum mattis ullamcorper velit ",
        },
        {
          "title": "ad minim veniam, qu",
          "description": "ue. Lectus vestibulum mattis ullamcorper velit ",
        },
        {
          "title": "ad minim veniam, qu",
          "description": "ue. Lectus vestibulum mattis ullamcorper velit ",
        },
        {
          "title": "ad minim veniam, qu",
          "description": "ue. Lectus vestibulum mattis ullamcorper velit ",
        }
      ],
      "2019-08-17": [
        {
          "title": "17th Aug",
          "description":
              "consectetur adipiscing elit, sed do eiusmod tempor incididunt ut",
        },
        {
          "title": "sed do eiusmod tempor ",
          "description":
              "consectetur adipiscing elit, sed do eiusmod tempor incididunt ut",
        }
      ],
      "2019-08-18": [
        {
          "title": "18th Aug",
          "description":
              "odo consequat. Duis aute irure dolor in reprehenderit in voluptate",
        },
        {
          "title": "incididunt ut labore",
          "description":
              "empor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam",
        }
      ],
      "2019-08-19": [
        {
          "title": "19th Aug",
          "description":
              "ermentum iaculis eu non diam phasellus. Massa tincidunt nunc "
        },
      ],
      "2019-08-20": [
        {
          "title": "20th Aug",
          "description":
              "ermentum iaculis eu non diam phasellus. Massa tincidunt nunc "
        },
      ],
      "2019-08-21": [
        {
          "title": "21 Aug",
          "description":
              "sed ullamcorper morbi tincidunt ornare. Feugiat vivamus at augue eget arcu. Arcu cursus euismod",
        },
        {
          "title": "ad minim veniam, qu",
          "description": "ue. Lectus vestibulum mattis ullamcorper velit ",
        },
        {
          "title": "ad minim veniam, qu",
          "description": "ue. Lectus vestibulum mattis ullamcorper velit ",
        },
        {
          "title": "ad minim veniam, qu",
          "description": "ue. Lectus vestibulum mattis ullamcorper velit ",
        },
        {
          "title": "ad minim veniam, qu",
          "description": "ue. Lectus vestibulum mattis ullamcorper velit ",
        }
      ],
      "2019-08-22": [
        {
          "title": "22 Aug",
          "description":
              "consectetur adipiscing elit, sed do eiusmod tempor incididunt ut",
        },
        {
          "title": "sed do eiusmod tempor ",
          "description":
              "consectetur adipiscing elit, sed do eiusmod tempor incididunt ut",
        }
      ],
      "2019-08-23": [
        {
          "title": "23 Aug",
          "description":
              "sed ullamcorper morbi tincidunt ornare. Feugiat vivamus at augue eget arcu. Arcu cursus euismod",
        },
        {
          "title": "ad minim veniam, qu",
          "description": "ue. Lectus vestibulum mattis ullamcorper velit ",
        },
        {
          "title": "ad minim veniam, qu",
          "description": "ue. Lectus vestibulum mattis ullamcorper velit ",
        },
        {
          "title": "ad minim veniam, qu",
          "description": "ue. Lectus vestibulum mattis ullamcorper velit ",
        },
        {
          "title": "ad minim veniam, qu",
          "description": "ue. Lectus vestibulum mattis ullamcorper velit ",
        }
      ],
      "2019-08-24": [
        {
          "title": "24 Aug",
          "description":
              "consectetur adipiscing elit, sed do eiusmod tempor incididunt ut",
        },
        {
          "title": "sed do eiusmod tempor ",
          "description":
              "consectetur adipiscing elit, sed do eiusmod tempor incididunt ut",
        }
      ],
      "2019-08-25": [
        {
          "title": "25 Aug",
          "description":
              "sed ullamcorper morbi tincidunt ornare. Feugiat vivamus at augue eget arcu. Arcu cursus euismod",
        },
        {
          "title": "ad minim veniam, qu",
          "description": "ue. Lectus vestibulum mattis ullamcorper velit ",
        },
        {
          "title": "ad minim veniam, qu",
          "description": "ue. Lectus vestibulum mattis ullamcorper velit ",
        },
        {
          "title": "ad minim veniam, qu",
          "description": "ue. Lectus vestibulum mattis ullamcorper velit ",
        },
        {
          "title": "ad minim veniam, qu",
          "description": "ue. Lectus vestibulum mattis ullamcorper velit ",
        }
      ],
    };
  }
}


test/scrolling_day_calendar_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:scrolling_day_calendar/scrolling_day_calendar.dart';

void main() {
  test('adds one to input values', () {
    final calculator = Calculator();
    expect(calculator.addOne(2), 3);
    expect(calculator.addOne(-7), -6);
    expect(calculator.addOne(0), 1);
    expect(() => calculator.addOne(null), throwsNoSuchMethodError);
  });
}


example/lib/page_two.dart
import 'package:flutter/material.dart';
import 'package:scrolling_day_calendar/scrolling_day_calendar.dart';
import 'package:example/test_data.dart';
import 'package:intl/intl.dart';
import 'package:example/page_one.dart';

class PageTwo extends StatefulWidget {
  @override
  _PageTwoState createState() => _PageTwoState();
}

class _PageTwoState extends State<PageTwo> {
  DateTime selectedDate = DateTime.now();
  DateTime startDate = DateTime.now().subtract(Duration(days: 10));
  DateTime endDate = DateTime.now().add(Duration(days: 10));
  String widgetKeyFormat = "yyyy-MM-dd";
  // test data
  Map testData = TestData.getTestData();

  Widget pageItems = Center(
    child: Text("No events"),
  );

  _widgetBuilder(DateTime selectedDate) {
    String dateString = DateFormat(widgetKeyFormat).format(selectedDate);

    // find the child records in test data and build widgets
    if (testData.containsKey(dateString)) {
      List items = testData[dateString];

      return ListView.builder(
        itemCount: items.length,
        itemBuilder: (context, key) {
          return Card(
            child: Padding(
              padding: EdgeInsets.all(10),
              child: Column(
                crossAxisAlignment: CrossAxisAlignment.start,
                mainAxisAlignment: MainAxisAlignment.start,
                children: <Widget>[
                  Text(
                    items[key]["title"],
                    style: TextStyle(fontWeight: FontWeight.bold),
                  ),
                  SizedBox(
                    height: 5.0,
                  ),
                  Text(items[key]["description"])
                ],
              ),
            ),
          );
        },
      );
    }

    // default widget to display on the page
    return Center(
      child: Text("No events"),
    );
  }

  @override
  Widget build(BuildContext context) {
    return SafeArea(
      child: Scaffold(
        drawer: Drawer(
          // Add a ListView to the drawer. This ensures the user can scroll
          // through the options in the drawer if there isn't enough vertical
          // space to fit everything.
          child: ListView(
            // Important: Remove any padding from the ListView.
            padding: EdgeInsets.zero,
            children: <Widget>[
              DrawerHeader(
                child: Text('Examples'),
                decoration: BoxDecoration(
                  color: Colors.blue,
                ),
              ),
              ListTile(
                title: Text('Pass-in Widgets Example'),
                onTap: () {
                  Navigator.push(
                    context,
                    MaterialPageRoute(builder: (context) => PageOne()),
                  );
                },
              ),
            ],
          ),
        ),
        appBar: AppBar(
          centerTitle: true,
          title: Text("SetState Example"),
        ),
        body: ScrollingDayCalendar(
          startDate: startDate,
          endDate: endDate,
          selectedDate: selectedDate,
          onDateChange: (direction, DateTime selectedDate) {
            setState(() {
              pageItems = _widgetBuilder(selectedDate);
            });
          },
          dateStyle: TextStyle(
            fontWeight: FontWeight.bold,
            color: Colors.white,
          ),
          pageItems: pageItems,
          displayDateFormat: "dd/MM/yyyy",
          dateBackgroundColor: Colors.grey,
          forwardIcon: Icons.arrow_forward,
          backwardIcon: Icons.arrow_back,
          pageChangeDuration: Duration(
            milliseconds: 400,
          ),
          noItemsWidget: Center(
            child: Text("No items have been added for this date"),
          ),
        ),
      ),
    );
  }
}


lib/scrolling_day_calendar.dart
library scrolling_day_calendar;

import 'package:flutter/material.dart';
import 'package:intl/intl.dart';

typedef ScrollingDayCalendarBuilder = Widget Function(
  BuildContext context,
  DateTime startDate,
  DateTime endDate,
  DateTime selectedDate,
  Function onDateChange,
);

class ScrollingDayCalendar extends StatefulWidget {
  // first date on the pages
  final DateTime startDate;
  // last date on the pages
  final DateTime endDate;
  // the active date
  final DateTime selectedDate;
  // what to do then the date changes
  final Function onDateChange;
  // page widgets to display
  final Widget pageItems;
  // date format
  final String displayDateFormat;
  // date style
  final TextStyle dateStyle;
  // background color for date container
  final Color dateBackgroundColor;
  // forward icon
  final IconData forwardIcon;
  // back icon
  final IconData backwardIcon;
  // page change duration
  final Duration pageChangeDuration;

  final Map<String, Widget> widgets;
  final Widget noItemsWidget;
  final String widgetKeyFormat;

  ScrollingDayCalendar({
    @required this.pageItems,
    @required this.startDate,
    @required this.endDate,
    @required this.selectedDate,
    this.onDateChange,
    this.widgets,
    this.noItemsWidget,
    this.widgetKeyFormat,
    this.displayDateFormat,
    this.dateStyle,
    this.dateBackgroundColor,
    this.forwardIcon,
    this.backwardIcon,
    this.pageChangeDuration,
  });

  @override
  _ScrollingDayCalendarState createState() => _ScrollingDayCalendarState();
}

class _ScrollingDayCalendarState extends State<ScrollingDayCalendar> {
  PageController _pageController;
  int _totalPages;
  int _currentPage;
  int _previousPage;
  DateTime _selectedDate;

  _onPageChange(direction) {
    _currentPage = _pageController.page.round();

    if (_currentPage > _previousPage) {
      // went forward
      DateTime newDate = _selectedDate.add(
        Duration(days: 1),
      );

      setState(() {
        _selectedDate = newDate;
      });
    } else {
      // went back
      DateTime newDate = _selectedDate.subtract(
        Duration(days: 1),
      );

      setState(() {
        _selectedDate = newDate;
      });
    }

    _previousPage = _pageController.page.round();

    // run page update sent by user
    if (widget.onDateChange != null) {
      widget.onDateChange(direction, _selectedDate);
    }
  }

  Widget _buildPage(index) {
    if (widget.pageItems != null) {
      return widget.pageItems;
    }
    DateTime dateTime = widget.startDate;
    index = index + 1;

    dateTime = widget.startDate.add(Duration(days: index));
    String key = DateFormat(widget.widgetKeyFormat).format(dateTime);

    if (widget.widgets != null && widget.widgets.containsKey(key)) {
      return widget.widgets[key];
    }

    return widget.noItemsWidget;
  }

  @override
  void dispose() {
    _pageController.dispose();
    super.dispose();
  }

  @override
  void initState() {
    // set the selected date
    _selectedDate = widget.selectedDate;

    // calculate the start page
    int startingPage =
        _selectedDate.difference(widget.startDate).inDays.floor();

    setState(() {
      // set the total number of pages based on start date and end date
      _totalPages = widget.endDate.difference(widget.startDate).inDays.floor();

      // set starting page
      _pageController = PageController(initialPage: startingPage);

      // set previous page
      _previousPage = startingPage;
    });

    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    return Column(
      children: <Widget>[
        Row(
          children: <Widget>[
            Container(
              height: 60.0,
              width: 60.0,
              color: widget.dateBackgroundColor != null
                  ? widget.dateBackgroundColor
                  : Colors.red,
              child: Center(
                child: MaterialButton(
                  onPressed: () {
                    _pageController.previousPage(
                      duration: widget.pageChangeDuration != null
                          ? widget.pageChangeDuration
                          : Duration(microseconds: 700),
                      curve: Curves.easeIn,
                    );
                  },
                  child: Icon(
                    widget.backwardIcon == null
                        ? Icons.arrow_back
                        : widget.backwardIcon,
                    color: Colors.white,
                  ),
                ),
              ),
            ),
            Expanded(
              child: Container(
                height: 60.0,
                color: widget.dateBackgroundColor != null
                    ? widget.dateBackgroundColor
                    : Colors.red,
                child: Padding(
                  padding: EdgeInsets.only(top: 15.0, bottom: 15.0),
                  child: Center(
                    child: Text(
                      DateFormat(widget.displayDateFormat != null
                              ? widget.displayDateFormat
                              : "dd/MM/yyyy")
                          .format(_selectedDate),
                      style: widget.dateStyle != null
                          ? widget.dateStyle
                          : TextStyle(
                              fontWeight: FontWeight.w500,
                              color: Colors.white,
                              fontSize: 18.0,
                            ),
                    ),
                  ),
                ),
              ),
            ),
            Container(
              height: 60.0,
              width: 60.0,
              color: widget.dateBackgroundColor != null
                  ? widget.dateBackgroundColor
                  : Colors.red,
              child: Center(
                child: MaterialButton(
                  onPressed: () {
                    _pageController.nextPage(
                      duration: widget.pageChangeDuration != null
                          ? widget.pageChangeDuration
                          : Duration(milliseconds: 700),
                      curve: Curves.easeIn,
                    );
                  },
                  child: Icon(
                    widget.forwardIcon == null
                        ? Icons.arrow_forward
                        : widget.forwardIcon,
                    color: Colors.white,
                  ),
                ),
              ),
            ),
          ],
        ),
        Expanded(
          child: PageView.builder(
            controller: _pageController,
            scrollDirection: Axis.horizontal,
            itemCount: _totalPages, // Can be null
            onPageChanged: (direction) => _onPageChange(direction),
            itemBuilder: (context, index) {
              return _buildPage(index);
            },
          ),
        ),
      ],
    );
  }
}


example/lib/page_one.dart
import 'package:flutter/material.dart';
import 'package:scrolling_day_calendar/scrolling_day_calendar.dart';
import 'package:example/test_data.dart';
import 'package:intl/intl.dart';
import 'package:example/page_two.dart';

class PageOne extends StatefulWidget {
  @override
  _PageOneState createState() => _PageOneState();
}

class _PageOneState extends State<PageOne> {
  DateTime selectedDate = DateTime.now();
  DateTime startDate = DateTime.now().subtract(Duration(days: 200));
  DateTime endDate = DateTime.now().add(Duration(days: 200));

  // this date will update every time the page changes
  DateTime _currentPageDate;

  // test data
  Map testData = TestData.getTestData();
  Map<String, Widget> widgets = Map();
  String widgetKeyFormat = "yyyy-MM-dd";

  _buildPages() {
    testData.forEach((key, values) {
      DateTime dateTime = DateTime.parse(key);
      var widget = _widgetBuilder(dateTime);
      widgets.addAll({key: widget});
    });

    return widgets;
  }

  _widgetBuilder(DateTime selectedDate) {
    String dateString = DateFormat(widgetKeyFormat).format(selectedDate);

    // find the child records in test data and build widgets
    if (testData.containsKey(dateString)) {
      List items = testData[dateString];

      return ListView.builder(
        itemCount: items.length,
        itemBuilder: (context, key) {
          return Card(
            child: Padding(
              padding: EdgeInsets.all(10),
              child: Column(
                crossAxisAlignment: CrossAxisAlignment.start,
                mainAxisAlignment: MainAxisAlignment.start,
                children: <Widget>[
                  Text(
                    items[key]["title"],
                    style: TextStyle(fontWeight: FontWeight.bold),
                  ),
                  SizedBox(
                    height: 5.0,
                  ),
                  Text(items[key]["description"])
                ],
              ),
            ),
          );
        },
      );
    }

    // default widget to display on the page
    return Center(
      child: Text("No events"),
    );
  }

  @override
  void initState() {
    _buildPages();
    _currentPageDate = selectedDate;
    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    return SafeArea(
      child: Scaffold(
        drawer: Drawer(
          // Add a ListView to the drawer. This ensures the user can scroll
          // through the options in the drawer if there isn't enough vertical
          // space to fit everything.
          child: ListView(
            // Important: Remove any padding from the ListView.
            padding: EdgeInsets.zero,
            children: <Widget>[
              DrawerHeader(
                child: Text('Examples'),
                decoration: BoxDecoration(
                  color: Colors.blue,
                ),
              ),
              ListTile(
                title: Text('SetState Example'),
                onTap: () {
                  Navigator.push(
                    context,
                    MaterialPageRoute(builder: (context) => PageTwo()),
                  );
                },
              ),
            ],
          ),
        ),
        appBar: AppBar(
          centerTitle: true,
          title: Text("Passing Widgets example"),
        ),
        body: ScrollingDayCalendar(
          startDate: startDate,
          endDate: endDate,
          selectedDate: selectedDate,
          dateStyle: TextStyle(
            fontWeight: FontWeight.bold,
            color: Colors.white,
          ),
          displayDateFormat: "dd, MMM yyyy",
          dateBackgroundColor: Colors.grey,
          forwardIcon: Icons.arrow_forward,
          backwardIcon: Icons.arrow_back,
          pageChangeDuration: Duration(
            milliseconds: 700,
          ),
          onDateChange: (direction, DateTime date) {
            _currentPageDate = date;
          },
          widgets: widgets,
          widgetKeyFormat: widgetKeyFormat,
          noItemsWidget: Center(
            child: MaterialButton(
              onPressed: () {
                String formattedDate =
                    DateFormat(widgetKeyFormat).format(_currentPageDate);

                testData[formattedDate] = [
                  {
                    "title": "NEW added 1 ",
                    "description":
                        "sed ullamcorper morbi tincidunt ornare. Feugiat vivamus at augue eget arcu. Arcu cursus euismod",
                  },
                  {
                    "title": "New Added 2",
                    "description":
                        "ue. Lectus vestibulum mattis ullamcorper velit ",
                  },
                ];

                setState(() {
                  _buildPages();
                });
              },
              child: Text("Add a new Item"),
            ),
          ),
        ),
      ),
    );
  }
}
